presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -569,64 +620,252 @@ private static final class SubfieldExtractor
             private final StandardFunctionResolution functionResolution;
             private final ExpressionOptimizer expressionOptimizer;
             private final ConnectorSession connectorSession;
-
-            private SubfieldExtractor(StandardFunctionResolution functionResolution, ExpressionOptimizer expressionOptimizer, ConnectorSession connectorSession)
+            private final FunctionAndTypeManager functionAndTypeManager;
+            private final boolean isPushDownSubfieldsFromLambdasEnabled;
+
+            private SubfieldExtractor(
+                    StandardFunctionResolution functionResolution,
+                    ExpressionOptimizer expressionOptimizer,
+                    ConnectorSession connectorSession,
+                    FunctionAndTypeManager functionAndTypeManager,
+                    boolean isPushDownSubfieldsFromLambdasEnabled)
             {
                 this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
                 this.expressionOptimizer = requireNonNull(expressionOptimizer, "expressionOptimizer is null");
-                this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+                this.connectorSession = connectorSession;
+                this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
+                ComplexTypeFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (isSubscriptOrElementAtFunction(call, functionResolution, functionAndTypeManager)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (subfield.isPresent()) {
+                        if (context.isPruningLambdaSubfieldsPossible()) {
+                            addRequiredLambdaSubfields(context, subfield.get());
+                        }
+                        else {
+                            context.subfields.add(subfield.get());
+                        }
+                    }
+                    else {
+                        call.getArguments().forEach(argument -> argument.accept(this, context));
+                    }
+                    return null;
+                }
+                if (!isPushDownSubfieldsFromLambdasEnabled) {
```

Comment:
Why do we check this only here? Shouldn't be check this at top? We may call functions above like addRequiredLambdaSubfields, when we want to disable this pushdown.

Reply:
We are calling it at the most top level.
Even if  `isPushDownSubfieldsFromLambdasEnabled=false` we still have to check`if (isSubscriptOrElementAtFunction(call, functionResolution, functionAndTypeManager))` branch.